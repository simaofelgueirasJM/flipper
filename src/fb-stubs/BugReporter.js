/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import LogManager from './Logger';

export default class BugReporter {
  constructor(logManager: LogManager) {}
  async report(title: string, body: string): Promise<number> {
    return Promise.resolve(-1);
  }
}
