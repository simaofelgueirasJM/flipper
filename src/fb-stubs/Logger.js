/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

export type LogTypes = 'error' | 'warn' | 'info' | 'debug';
export type TrackType = 'duration' | 'usage' | 'performance';
import ScribeLogger from './ScribeLogger';

export default class LogManager {
  constructor() {
    this.scribeLogger = new ScribeLogger(this);
  }

  scribeLogger: ScribeLogger;

  track(type: TrackType, event: string, data: ?any) {}

  trackTimeSince(mark: string, eventName: ?string) {}

  info(data: any, category: string) {}

  warn(data: any, category: string) {}

  error(data: any, category: string) {}

  debug(data: any, category: string) {}
}
