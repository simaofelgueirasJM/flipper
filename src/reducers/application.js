/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import {remote} from 'electron';

export type State = {
  leftSidebarVisible: boolean,
  rightSidebarVisible: boolean,
  rightSidebarAvailable: boolean,
  bugDialogVisible: boolean,
  windowIsFocused: boolean,
  pluginManagerVisible: boolean,
};

type ActionType =
  | 'leftSidebarVisible'
  | 'rightSidebarVisible'
  | 'rightSidebarAvailable'
  | 'bugDialogVisible'
  | 'windowIsFocused'
  | 'pluginManagerVisible';

export type Action = {
  type: ActionType,
  payload?: boolean,
};

const initialState: () => State = () => ({
  leftSidebarVisible: true,
  rightSidebarVisible: true,
  rightSidebarAvailable: false,
  bugDialogVisible: false,
  windowIsFocused: remote.getCurrentWindow().isFocused(),
  pluginManagerVisible: false,
});

export default function reducer(state: State, action: Action): State {
  state = state || initialState();
  const {payload, type} = action;
  const newValue = typeof payload === 'undefined' ? !state[type] : payload;

  if (
    type === 'leftSidebarVisible' ||
    type === 'rightSidebarVisible' ||
    type === 'rightSidebarAvailable' ||
    type === 'bugDialogVisible' ||
    type === 'windowIsFocused' ||
    type === 'pluginManagerVisible'
  ) {
    if (state[type] === newValue) {
      // value hasn't changed
      return state;
    } else {
      return {
        ...state,
        [type]: newValue,
      };
    }
  } else {
    return state;
  }
}

export const toggleAction = (type: ActionType, payload?: boolean): Action => ({
  type,
  payload,
});

export const toggleBugDialogVisible = (payload?: boolean): Action => ({
  type: 'bugDialogVisible',
  payload,
});

export const toggleLeftSidebarVisible = (payload?: boolean): Action => ({
  type: 'leftSidebarVisible',
  payload,
});

export const toggleRightSidebarVisible = (payload?: boolean): Action => ({
  type: 'rightSidebarVisible',
  payload,
});

export const toggleRightSidebarAvailable = (payload?: boolean): Action => ({
  type: 'rightSidebarAvailable',
  payload,
});

export const togglePluginManagerVisible = (payload?: boolean): Action => ({
  type: 'pluginManagerVisible',
  payload,
});
