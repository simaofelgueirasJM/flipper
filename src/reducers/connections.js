/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import type BaseDevice from '../devices/BaseDevice';
import type Client from '../Client';

export type State = {|
  devices: Array<BaseDevice>,
  androidEmulators: Array<string>,
  selectedDevice: ?BaseDevice,
  selectedPlugin: ?string,
  selectedApp: ?string,
  userPreferredDevice: ?string,
  userPreferredPlugin: ?string,
  userPreferredApp: ?string,
  error: ?string,
  clients: Array<Client>,
  deepLinkPayload: ?string,
|};

export type Action =
  | {
      type: 'UNREGISTER_DEVICES',
      payload: Set<string>,
    }
  | {
      type: 'REGISTER_DEVICE',
      payload: BaseDevice,
    }
  | {
      type: 'REGISTER_ANDROID_EMULATORS',
      payload: Array<string>,
    }
  | {
      type: 'SELECT_DEVICE',
      payload: BaseDevice,
    }
  | {
      type: 'SELECT_PLUGIN',
      payload: {|
        selectedPlugin: ?string,
        selectedApp: ?string,
        deepLinkPayload: ?string,
      |},
    }
  | {
      type: 'SELECT_USER_PREFERRED_PLUGIN',
      payload: string,
    }
  | {
      type: 'SERVER_ERROR',
      payload: ?string,
    }
  | {
      type: 'NEW_CLIENT',
      payload: Client,
    }
  | {
      type: 'CLIENT_REMOVED',
      payload: string,
    }
  | {
      type: 'PREFER_DEVICE',
      payload: string,
    };

const DEFAULT_PLUGIN = 'DeviceLogs';

const INITAL_STATE: State = {
  devices: [],
  androidEmulators: [],
  selectedDevice: null,
  selectedApp: null,
  selectedPlugin: DEFAULT_PLUGIN,
  userPreferredDevice: null,
  userPreferredPlugin: null,
  userPreferredApp: null,
  error: null,
  clients: [],
  deepLinkPayload: null,
};

export default function reducer(
  state: State = INITAL_STATE,
  action: Action,
): State {
  switch (action.type) {
    case 'SELECT_DEVICE': {
      const {payload} = action;
      return {
        ...state,
        selectedApp: null,
        selectedPlugin: DEFAULT_PLUGIN,
        selectedDevice: payload,
        userPreferredDevice: payload.title,
      };
    }
    case 'REGISTER_ANDROID_EMULATORS': {
      const {payload} = action;
      return {
        ...state,
        androidEmulators: payload,
      };
    }
    case 'REGISTER_DEVICE': {
      const {payload} = action;
      const devices = state.devices.concat(payload);
      let {selectedDevice, selectedPlugin} = state;

      // select the default plugin
      let selection = {
        selectedApp: null,
        selectedPlugin: DEFAULT_PLUGIN,
      };

      if (!selectedDevice) {
        selectedDevice = payload;
        if (selectedPlugin) {
          // We already had a plugin selected, but no device. This is happening
          // when the Client connected before the Device.
          selection = {};
        }
      } else if (payload.title === state.userPreferredDevice) {
        selectedDevice = payload;
      } else {
        // We didn't select the newly connected device, so we don't want to
        // change the plugin.
        selection = {};
      }

      return {
        ...state,
        devices,
        // select device if none was selected before
        selectedDevice,
        ...selection,
      };
    }
    case 'UNREGISTER_DEVICES': {
      const {payload} = action;
      const {selectedDevice} = state;
      let selectedDeviceWasRemoved = false;
      const devices = state.devices.filter((device: BaseDevice) => {
        if (payload.has(device.serial)) {
          if (selectedDevice === device) {
            // removed device is the selected
            selectedDeviceWasRemoved = true;
          }
          return false;
        } else {
          return true;
        }
      });

      let selection = {};
      if (selectedDeviceWasRemoved) {
        selection = {
          selectedDevice: devices[devices.length - 1],
          selectedApp: null,
          selectedPlugin: DEFAULT_PLUGIN,
        };
      }

      return {
        ...state,
        devices,
        ...selection,
      };
    }
    case 'SELECT_PLUGIN': {
      const {payload} = action;
      const {selectedPlugin, selectedApp} = payload;
      if (selectedPlugin) {
        performance.mark(`activePlugin-${selectedPlugin}`);
      }

      return {
        ...state,
        ...payload,
        userPreferredApp: selectedApp || state.userPreferredApp,
        userPreferredPlugin: selectedPlugin,
      };
    }
    case 'SELECT_USER_PREFERRED_PLUGIN': {
      const {payload} = action;
      return {...state, userPreferredPlugin: payload};
    }
    case 'NEW_CLIENT': {
      const {payload} = action;
      const {userPreferredApp, userPreferredPlugin} = state;
      let {selectedApp, selectedPlugin} = state;

      if (
        userPreferredApp &&
        userPreferredPlugin &&
        payload.id === userPreferredApp &&
        payload.plugins.includes(userPreferredPlugin)
      ) {
        // user preferred client did reconnect, so let's select it
        selectedApp = userPreferredApp;
        selectedPlugin = userPreferredPlugin;
      }

      return {
        ...state,
        clients: state.clients.concat(payload),
        selectedApp,
        selectedPlugin,
      };
    }
    case 'CLIENT_REMOVED': {
      const {payload} = action;

      let selected = {};
      if (state.selectedApp === payload) {
        selected.selectedApp = null;
        selected.selectedPlugin = DEFAULT_PLUGIN;
      }

      return {
        ...state,
        ...selected,
        clients: state.clients.filter(
          (client: Client) => client.id !== payload,
        ),
      };
    }
    case 'PREFER_DEVICE': {
      const {payload: userPreferredDevice} = action;
      return {...state, userPreferredDevice};
    }
    case 'SERVER_ERROR': {
      const {payload} = action;
      return {...state, error: payload};
    }
    default:
      return state;
  }
}

export const selectDevice = (payload: BaseDevice): Action => ({
  type: 'SELECT_DEVICE',
  payload,
});

export const preferDevice = (payload: string): Action => ({
  type: 'PREFER_DEVICE',
  payload,
});

export const selectPlugin = (payload: {|
  selectedPlugin: ?string,
  selectedApp?: ?string,
  deepLinkPayload: ?string,
|}): Action => ({
  type: 'SELECT_PLUGIN',
  payload,
});
export const userPreferredPlugin = (payload: string): Action => ({
  type: 'SELECT_USER_PREFERRED_PLUGIN',
  payload,
});
