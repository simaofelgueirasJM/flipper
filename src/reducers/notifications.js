/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */
import type {Notification} from '../plugin';

export type PluginNotification = {|
  notification: Notification,
  pluginId: string,
  client: ?string,
|};

export type State = {
  activeNotifications: Array<PluginNotification>,
  invalidatedNotifications: Array<PluginNotification>,
  blacklistedPlugins: Array<string>,
  clearedNotifications: Set<string>,
};

type ActiveNotificationsAction = {
  type: 'SET_ACTIVE_NOTIFICATIONS',
  payload: {
    notifications: Array<Notification>,
    client: ?string,
    pluginId: string,
  },
};

type ClearAllAction = {
  type: 'CLEAR_ALL_NOTIFICATIONS',
};

type UpdateBlacklistAction = {
  type: 'UPDATE_PLUGIN_BLACKLIST',
  payload: Array<string>,
};

export type Action =
  | ActiveNotificationsAction
  | ClearAllAction
  | UpdateBlacklistAction;

const INITIAL_STATE: State = {
  activeNotifications: [],
  invalidatedNotifications: [],
  blacklistedPlugins: [],
  clearedNotifications: new Set(),
};

export default function reducer(
  state: State = INITIAL_STATE,
  action: Action,
): State {
  switch (action.type) {
    case 'SET_ACTIVE_NOTIFICATIONS': {
      return activeNotificationsReducer(state, action);
    }
    case 'CLEAR_ALL_NOTIFICATIONS':
      const markAsCleared = ({
        pluginId,
        notification: {id},
      }: PluginNotification) =>
        state.clearedNotifications.add(`${pluginId}#${id}`);

      state.activeNotifications.forEach(markAsCleared);
      state.invalidatedNotifications.forEach(markAsCleared);
      // Q: Should this actually delete them, or just invalidate them?
      return {
        ...state,
        activeNotifications: [],
        invalidatedNotifications: [],
      };
    case 'UPDATE_PLUGIN_BLACKLIST':
      return {
        ...state,
        blacklistedPlugins: action.payload,
      };
    default:
      return state;
  }
}

function activeNotificationsReducer(
  state: State,
  action: ActiveNotificationsAction,
): State {
  const {payload} = action;
  const newActiveNotifications = [];
  const newInactivatedNotifications = state.invalidatedNotifications;

  const newIDs = new Set(payload.notifications.map((n: Notification) => n.id));

  for (const activeNotification of state.activeNotifications) {
    if (activeNotification.pluginId !== payload.pluginId) {
      newActiveNotifications.push(activeNotification);
      continue;
    }

    if (!newIDs.has(activeNotification.notification.id)) {
      newInactivatedNotifications.push(activeNotification);
    }
  }

  payload.notifications
    .filter(
      ({id}: Notification) =>
        !state.clearedNotifications.has(`${payload.pluginId}#${id}`),
    )
    .forEach((notification: Notification) => {
      newActiveNotifications.push({
        pluginId: payload.pluginId,
        client: payload.client,
        notification,
      });
    });

  return {
    ...state,
    activeNotifications: newActiveNotifications,
    invalidatedNotifications: newInactivatedNotifications,
  };
}

export function setActiveNotifications(payload: {
  notifications: Array<Notification>,
  client: ?string,
  pluginId: string,
}): Action {
  return {
    type: 'SET_ACTIVE_NOTIFICATIONS',
    payload,
  };
}

export function clearAllNotifications(): Action {
  return {
    type: 'CLEAR_ALL_NOTIFICATIONS',
  };
}

export function updatePluginBlacklist(
  payload: Array<string>,
): UpdateBlacklistAction {
  return {
    type: 'UPDATE_PLUGIN_BLACKLIST',
    payload,
  };
}
