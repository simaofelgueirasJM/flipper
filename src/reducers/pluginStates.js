/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

export type State = {
  [pluginKey: string]: Object,
};

export type Action =
  | {
      type: 'SET_PLUGIN_STATE',
      payload: {
        pluginKey: string,
        state: Object,
      },
    }
  | {
      type: 'CLEAR_PLUGIN_STATE',
      payload: string,
    };

const INITIAL_STATE: State = {};

export default function reducer(
  state: State = INITIAL_STATE,
  action: Action,
): State {
  if (action.type === 'SET_PLUGIN_STATE') {
    return {
      ...state,
      [action.payload.pluginKey]: {
        ...state[action.payload.pluginKey],
        ...action.payload.state,
      },
    };
  } else if (action.type === 'CLEAR_PLUGIN_STATE') {
    const {payload} = action;
    return Object.keys(state).reduce((newState, pluginKey) => {
      // Only add the pluginState, if its from a plugin other than the one that
      // was removed. pluginKeys are in the form of ${clientID}#${pluginID}.
      if (!pluginKey.startsWith(payload)) {
        newState[pluginKey] = state[pluginKey];
      }
      return newState;
    }, {});
  } else {
    return state;
  }
}

export const setPluginState = (payload: {
  pluginKey: string,
  state: Object,
}): Action => ({
  type: 'SET_PLUGIN_STATE',
  payload,
});
