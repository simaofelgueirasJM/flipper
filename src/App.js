/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import React from 'react';
import {FlexColumn, FlexRow} from 'flipper';
import {connect} from 'react-redux';
import {toggleBugDialogVisible} from './reducers/application.js';
import WelcomeScreen from './chrome/WelcomeScreen.js';
import TitleBar from './chrome/TitleBar.js';
import MainSidebar from './chrome/MainSidebar.js';
import BugReporterDialog from './chrome/BugReporterDialog.js';
import ErrorBar from './chrome/ErrorBar.js';
import PluginContainer from './PluginContainer.js';
import {ipcRenderer} from 'electron';

import type Logger from './fb-stubs/Logger.js';
import type BugReporter from './fb-stubs/BugReporter.js';
import type BaseDevice from './devices/BaseDevice.js';

type Props = {
  logger: Logger,
  bugReporter: BugReporter,
  leftSidebarVisible: boolean,
  bugDialogVisible: boolean,
  pluginManagerVisible: boolean,
  selectedDevice: ?BaseDevice,
  error: ?string,
  toggleBugDialogVisible: (visible?: boolean) => any,
};

export class App extends React.Component<Props> {
  componentDidMount() {
    // track time since launch
    const [s, ns] = process.hrtime();
    const launchEndTime = s * 1e3 + ns / 1e6;
    ipcRenderer.on('getLaunchTime', (event, launchStartTime) => {
      this.props.logger.track(
        'performance',
        'launchTime',
        launchEndTime - launchStartTime,
      );
    });
    ipcRenderer.send('getLaunchTime');
    ipcRenderer.send('componentDidMount');
  }
  render() {
    return (
      <FlexColumn grow={true}>
        <TitleBar />
        {this.props.bugDialogVisible && (
          <BugReporterDialog
            bugReporter={this.props.bugReporter}
            close={() => {
              this.props.toggleBugDialogVisible(false);
            }}
          />
        )}
        <FlexRow grow={true}>
          {this.props.leftSidebarVisible && <MainSidebar />}
          {this.props.selectedDevice ? (
            <PluginContainer logger={this.props.logger} />
          ) : (
            <WelcomeScreen />
          )}
        </FlexRow>
        <ErrorBar text={this.props.error} />
      </FlexColumn>
    );
  }
}
export default connect(
  ({
    application: {pluginManagerVisible, bugDialogVisible, leftSidebarVisible},
    connections: {selectedDevice, error},
  }) => ({
    pluginManagerVisible,
    bugDialogVisible,
    leftSidebarVisible,
    selectedDevice,
    error,
  }),
  {
    toggleBugDialogVisible,
  },
)(App);
