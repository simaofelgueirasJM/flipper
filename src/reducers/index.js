/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import {combineReducers} from 'redux';
import application from './application.js';
import connections from './connections.js';
import pluginStates from './pluginStates.js';
import notifications from './notifications.js';
import {persistReducer} from 'redux-persist';
import storage from 'redux-persist/lib/storage/index.js';

import type {
  State as ApplicationState,
  Action as ApplicationAction,
} from './application.js';
import type {
  State as DevicesState,
  Action as DevicesAction,
} from './connections.js';
import type {
  State as PluginsState,
  Action as PluginsAction,
} from './pluginStates.js';
import type {
  State as NotificationsState,
  Action as NotificationsAction,
} from './notifications.js';
import type {Store as ReduxStore} from 'redux';

export type Store = ReduxStore<
  {
    application: ApplicationState,
    connections: DevicesState,
    pluginStates: PluginsState,
    notifications: NotificationsState,
  },
  | ApplicationAction
  | DevicesAction
  | PluginsAction
  | NotificationsAction
  | {|type: 'INIT'|},
>;

export default combineReducers({
  application,
  connections: persistReducer(
    {
      key: 'connections',
      storage,
      whitelist: [
        'userPreferredDevice',
        'userPreferredPlugin',
        'userPreferredApp',
      ],
    },
    connections,
  ),
  pluginStates,
  notifications,
});
