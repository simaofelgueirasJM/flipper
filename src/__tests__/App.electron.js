/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */
import React from 'react';
import {App} from '../App.js';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import reducers from '../reducers/index.js';
import configureStore from 'redux-mock-store';
import Logger from '../fb-stubs/Logger.js';
import BugReporter from '../fb-stubs/BugReporter.js';

// create redux store with initial state
const mockStore = configureStore([])(reducers(undefined, {type: 'INIT'}));

test('Empty app state matches snapshot', () => {
  const logger = new Logger();
  const bugReporter = new BugReporter(logger);

  const component = renderer.create(
    <Provider store={mockStore}>
      <App
        logger={logger}
        bugReporter={bugReporter}
        leftSidebarVisible={false}
        bugDialogVisible={false}
        pluginManagerVisible={false}
        selectedDevice={null}
        toggleBugDialogVisible={() => {}}
        error={null}
      />
    </Provider>,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
