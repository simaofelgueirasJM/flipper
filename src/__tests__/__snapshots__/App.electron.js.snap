// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty app state matches snapshot 1`] = `
<div
  className="css-1si6n3e"
>
  <div
    className="toolbar css-1u64wvw"
  >
    <div
      className="css-7nzs1f"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="css-at56xe"
        color="#acacac"
        size={12}
        src="https://external.xx.fbcdn.net/assets/?name=minus-circle&variant=filled&size=12&set=facebook_icons&density=1x"
      />
      No device selected
    </div>
    <div
      className="css-1ayt83l"
    >
      <div
        className="css-1telx8z"
        disabled={true}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Take Screenshot"
      >
        <div
          className="css-11ecny9"
          color="#acacac"
          size={12}
          src="https://external.xx.fbcdn.net/assets/?name=camera&variant=filled&size=12&set=facebook_icons&density=1x"
        />
      </div>
      <div
        className="css-1telx8z"
        disabled={true}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Make Screen Recording"
      >
        <div
          className="css-1qxark3"
          color="#acacac"
          size={12}
          src="https://external.xx.fbcdn.net/assets/?name=camcorder&variant=filled&size=12&set=facebook_icons&density=1x"
        />
      </div>
    </div>
    <div
      className="css-12zzrdt"
    />
    <div
      className="css-1xcv92o"
    />
    <div
      className="css-1lljw3m"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      title="Report Bug"
    >
      <div
        className="css-1u10tgp"
        color="#acacac"
        size={12}
        src="https://external.xx.fbcdn.net/assets/?name=bug&variant=filled&size=12&set=facebook_icons&density=1x"
      />
    </div>
    <div
      className="css-1ayt83l"
    >
      <div
        className="css-19m5dks"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Toggle Plugins"
      >
        <div
          className="css-1pfn0zl"
          color="#80a6f5"
          size={20}
          src="icons/sidebar_left.svg"
        />
      </div>
      <div
        className="css-1telx8z"
        disabled={true}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Toggle Details"
      >
        <div
          className="css-13cmaba"
          color="#acacac"
          size={20}
          src="icons/sidebar_right.svg"
        />
      </div>
    </div>
  </div>
  <div
    className="css-9qtipk"
  >
    <div
      className="css-1atbcdi"
    >
      <div
        className="css-1hs1k35"
      >
        <img
          className="css-1yl4k4p"
          src="./icon.png"
        />
        <span
          className="css-p2pwny"
        >
          Welcome to Flipper
        </span>
        <span
          className="css-1xkgddx"
        >
          Development Mode
        </span>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-1wavjrl"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=rocket&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Using Flipper
            </span>
            <span
              className="css-bcc6f5"
            >
              Learn how Flipper can help you debug your App
            </span>
          </div>
        </div>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-1bmw9ne"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=magic-wand&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Create your own plugin
            </span>
            <span
              className="css-bcc6f5"
            >
              Get started with these pointers
            </span>
          </div>
        </div>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-nk9hkx"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=tools&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Add Flipper support to your app
            </span>
            <span
              className="css-bcc6f5"
            >
              Get started with these pointers
            </span>
          </div>
        </div>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-14xjmfa"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=posts&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Contributing and Feedback
            </span>
            <span
              className="css-bcc6f5"
            >
              Report issues and help us improve Flipper
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
